File: apps/openai-agents-worker/tools.mjs
// Static trusted functions only. Requests select IDs; they never supply code,
// schemas, network destinations, filesystem paths, or command arguments.
const definitions = Object.freeze({
  calculator: Object.freeze({ name: 'calculator', description: 'Calculate bounded arithmetic with decimal numbers, + - * / and parentheses.',
    parameters: { type: 'object', properties: { expression: { type: 'string', maxLength: 256 } }, required: ['expression'], additionalProperties: false } }),
  current_time: Object.freeze({ name: 'current_time', description: 'Return the current ISO UTC timestamp and optionally a validated IANA timezone display.',
    parameters: { type: 'object', properties: { timeZone: { type: ['string', 'null'], maxLength: 80 } }, required: ['timeZone'], additionalProperties: false } }),
});
export function validateToolNames(names) {
  if (!Array.isArray(names) || names.length > 2 || new Set(names).size !== names.length
    || names.some(name => typeof name !== 'string' || !Object.hasOwn(definitions, name))) throw new Error('Invalid tool selection');
  return [...names];
}
export const toolMetadata = name => ({ id: name, name, type: 'function', readOnly: true, description: definitions[name].description, contextTextBytes: toolDefinitionBytes([name]) });
export const toolDefinitionBytes = names => names.reduce((bytes, name) => bytes + Buffer.byteLength(JSON.stringify(definitions[name])) + 128, 0);

export class ToolInputError extends Error { constructor() { super('Invalid registered tool input'); this.name = 'ToolInputError'; } }
export function parseToolArguments(name, serialized) {
  let input;
  try { input = JSON.parse(serialized); } catch { throw new ToolInputError(); }
  if (!input || typeof input !== 'object' || Array.isArray(input) || Object.keys(input).length !== 1) throw new ToolInputError();
  if (name === 'calculator' && Object.hasOwn(input, 'expression') && typeof input.expression === 'string') return input;
  if (name === 'current_time' && Object.hasOwn(input, 'timeZone') && (input.timeZone === null || typeof input.timeZone === 'string')) return input;
  throw new ToolInputError();
}
function bounded(value) { if (!Number.isFinite(value) || Math.abs(value) > 1e12) throw new ToolInputError(); return value; }
export function calculate(expression) {
  if (typeof expression !== 'string' || !expression.length || expression.length > 256 || /[^0-9. +*/()\-\t]/.test(expression)) throw new ToolInputError();
  const tokens = expression.match(/(?:\d+(?:\.\d*)?|\.\d+)|[()+*/-]/g) ?? [];
  if (tokens.join('') !== expression.replace(/[ \t]/g, '') || tokens.length > 128) throw new ToolInputError();
  let at = 0;
  function atom(depth) {
    if (depth > 16) throw new ToolInputError();
    const token = tokens[at++];
    if (token === '+' || token === '-') return bounded((token === '-' ? -1 : 1) * atom(depth + 1));
    if (token === '(') { const value = sum(depth + 1); if (tokens[at++] !== ')') throw new ToolInputError(); return value; }
    if (!token || !/^(?:\d+(?:\.\d*)?|\.\d+)$/.test(token)) throw new ToolInputError();
    return bounded(Number(token));
  }
  function product(depth) {
    let value = atom(depth);
    while (['*', '/'].includes(tokens[at])) { const op = tokens[at++], rhs = atom(depth); value = bounded(op === '*' ? value * rhs : value / rhs); }
    return value;
  }
  function sum(depth) {
    let value = product(depth);
    while (['+', '-'].includes(tokens[at])) { const op = tokens[at++], rhs = product(depth); value = bounded(op === '+' ? value + rhs : value - rhs); }
    return value;
  }
  const value = sum(0); if (at !== tokens.length) throw new ToolInputError(); return Object.is(value, -0) ? 0 : value;
}
export function executeTool(name, input, now = () => new Date()) {
  if (!input || typeof input !== 'object' || Array.isArray(input)) throw new ToolInputError();
  if (name === 'calculator' && Object.keys(input).length === 1 && Object.hasOwn(input, 'expression')) return JSON.stringify({ value: calculate(input.expression) });
  if (name === 'current_time' && Object.keys(input).length === 1 && Object.hasOwn(input, 'timeZone')) {
    const timeZone = input.timeZone ?? 'UTC';
    if (typeof timeZone !== 'string' || timeZone.length > 80 || !/^[A-Za-z_]+(?:\/[A-Za-z0-9_+\-]+){0,2}$/.test(timeZone)) throw new ToolInputError();
    try {
      const date = now();
      const local = new Intl.DateTimeFormat('en-CA', { timeZone, dateStyle: 'short', timeStyle: 'long' }).format(date);
      return JSON.stringify({ iso: date.toISOString(), timeZone, local });
    } catch { throw new ToolInputError(); }
  }
  throw new ToolInputError();
}
export function registeredTools(names, tool, { signal, onTool } = {}) {
  return validateToolNames(names).map(name => tool({ ...definitions[name], strict: true, errorFunction: null,
    execute: input => {
      if (signal?.aborted) throw new DOMException('Cancelled', 'AbortError');
      const output = executeTool(name, input);
      onTool?.({ name }); // Metadata only; no arguments, clock output or user data in diagnostics.
      return output;
    },
  }));
}

File: apps/openai-agents-python-worker/tools.py
"""Static trusted tools, mirroring the Node worker's tools.mjs.

Requests select tool IDs; they never supply code, schemas, network
destinations, filesystem paths, or command arguments.
"""

import json
import re
from datetime import datetime, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError


class ToolInputError(ValueError):
    pass


def _bounded(value: float) -> float:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise ToolInputError()
    if value != value or value in (float("inf"), float("-inf")) or abs(value) > 1e12:
        raise ToolInputError()
    return float(value)


def calculate(expression: str) -> float:
    if not isinstance(expression, str) or not expression or len(expression) > 256:
        raise ToolInputError()
    if re.search(r"[^0-9. +*/()\-\t]", expression):
        raise ToolInputError()
    tokens = re.findall(r"(?:\d+(?:\.\d*)?|\.\d+)|[()+*/-]", expression)
    if "".join(tokens) != re.sub(r"[ \t]", "", expression) or len(tokens) > 128:
        raise ToolInputError()

    state = {"at": 0}

    def atom(depth: int) -> float:
        if depth > 16:
            raise ToolInputError()
        token = tokens[state["at"]] if state["at"] < len(tokens) else None
        state["at"] += 1
        if token == "+":
            return atom(depth + 1)
        if token == "-":
            return -atom(depth + 1)
        if token == "(":
            value = sum_expr(depth + 1)
            if state["at"] >= len(tokens) or tokens[state["at"]] != ")":
                raise ToolInputError()
            state["at"] += 1
            return value
        if not token or not re.fullmatch(r"(?:\d+(?:\.\d*)?|\.\d+)", token):
            raise ToolInputError()
        return _bounded(float(token))

    def product(depth: int) -> float:
        value = atom(depth)
        while state["at"] < len(tokens) and tokens[state["at"]] in ("*", "/"):
            op = tokens[state["at"]]
            state["at"] += 1
            rhs = atom(depth)
            value = value * rhs if op == "*" else value / rhs
        return value

    def sum_expr(depth: int) -> float:
        value = product(depth)
        while state["at"] < len(tokens) and tokens[state["at"]] in ("+", "-"):
            op = tokens[state["at"]]
            state["at"] += 1
            rhs = product(depth)
            value = value + rhs if op == "+" else value - rhs
        return value

    value = sum_expr(0)
    if state["at"] != len(tokens):
        raise ToolInputError()
    return 0.0 if value == 0 else value


def execute_tool(name: str, tool_input: dict) -> str:
    if not isinstance(tool_input, dict):
        raise ToolInputError()
    if name == "calculator":
        if list(tool_input.keys()) != ["expression"]:
            raise ToolInputError()
        return json.dumps({"value": calculate(tool_input["expression"])})
    if name == "current_time":
        if list(tool_input.keys()) != ["timeZone"]:
            raise ToolInputError()
        time_zone = tool_input["timeZone"] or "UTC"
        if not isinstance(time_zone, str) or len(time_zone) > 80:
            raise ToolInputError()
        if not re.fullmatch(r"[A-Za-z_]+(?:/[A-Za-z0-9_+-]+){0,2}", time_zone):
            raise ToolInputError()
        try:
            tz = ZoneInfo(time_zone)
        except (ZoneInfoNotFoundError, ValueError):
            raise ToolInputError()
        now = datetime.now(timezone.utc)
        return json.dumps({"iso": now.isoformat(), "timeZone": time_zone, "local": now.astimezone(tz).isoformat()})
    raise ToolInputError()


def tool_definition_bytes(names: list[str]) -> int:
    total = 0
    for name in names:
        schema = TOOL_SCHEMAS[name]
        total += len(json.dumps(schema).encode("utf-8")) + 128
    return total


TOOL_SCHEMAS = {
    "calculator": {
        "name": "calculator",
        "description": "Calculate bounded arithmetic with decimal numbers, + - * / and parentheses.",
        "parameters": {
            "type": "object",
            "properties": {"expression": {"type": "string", "maxLength": 256}},
            "required": ["expression"],
            "additionalProperties": False,
        },
    },
    "current_time": {
        "name": "current_time",
        "description": "Return the current ISO UTC timestamp and optionally a validated IANA timezone display.",
        "parameters": {
            "type": "object",
            "properties": {"timeZone": {"type": ["string", "null"], "maxLength": 80}},
            "required": ["timeZone"],
            "additionalProperties": False,
        },
    },
}


def validate_tool_names(names) -> list[str]:
    if not isinstance(names, list) or len(names) > 2 or len(set(names)) != len(names):
        raise ToolInputError()
    for name in names:
        if not isinstance(name, str) or name not in TOOL_SCHEMAS:
            raise ToolInputError()
    return list(names)


def tool_metadata(name: str) -> dict:
    return {
        "id": name,
        "name": name,
        "type": "function",
        "readOnly": True,
        "description": TOOL_SCHEMAS[name]["description"],
        "contextTextBytes": tool_definition_bytes([name]),
    }
